com.unity.ml-agents/Runtime/Actuators/ActuatorDiscreteActionMask.cs

```diff
@@ -0,0 +1,151 @@
+
+        int[] m_BranchSizes;
+
+        bool[] m_CurrentMask;
+
+        IList<IActuator> m_Actuators;
+
+        readonly int m_SumOfDiscreteBranchSizes;
+        readonly int m_NumBranches;
+
+        /// <summary>
+        /// The offset into the branches array that is used when actuators are writing to the action mask.
+        /// </summary>
+        public int CurrentBranchOffset { get; set; }
+
+        internal ActuatorDiscreteActionMask(IList<IActuator> actuators, int sumOfDiscreteBranchSizes, int numBranches)
+        {
+            m_Actuators = actuators;
+            m_SumOfDiscreteBranchSizes = sumOfDiscreteBranchSizes;
+            m_NumBranches = numBranches;
+        }
+
+        /// <inheritdoc cref="IDiscreteActionMask.WriteMask"/>
+        public void WriteMask(int branch, IEnumerable<int> actionIndices)
+        {
+            LazyInitialize();
+
+            // Perform the masking
+            foreach (var actionIndex in actionIndices)
+            {
+#if DEBUG
+                if (branch >= m_NumBranches || actionIndex >= m_BranchSizes[CurrentBranchOffset + branch])
+                {
+                    throw new UnityAgentsException(
+                        "Invalid Action Masking: Action Mask is too large for specified branch.");
+                }
+#endif
+                m_CurrentMask[actionIndex + m_StartingActionIndices[CurrentBranchOffset + branch]] = true;
+            }
+        }
+
+        void LazyInitialize()
+        {
+            if (m_BranchSizes == null)
+            {
+                m_BranchSizes = new int[m_NumBranches];
+                var start = 0;
+                for (var i = 0; i < m_Actuators.Count; i++)
+                {
+                    var actuator = m_Actuators[i];
+                    var branchSizes = actuator.actionSpecs.BranchSizes;
+                    Array.Copy(branchSizes, 0, m_BranchSizes, start, branchSizes.Length);
+                    start += branchSizes.Length;
+                }
+            }
+
+            // By default, the masks are null. If we want to specify a new mask, we initialize
+            // the actionMasks with trues.
+            if (m_CurrentMask == null)
+            {
+                m_CurrentMask = new bool[m_SumOfDiscreteBranchSizes];
+            }
+
+            // If this is the first time the masked actions are used, we generate the starting
+            // indices for each branch.
+            if (m_StartingActionIndices == null)
+            {
+                m_StartingActionIndices = Utilities.CumSum(m_BranchSizes);
+            }
+        }
+
+        /// <inheritdoc cref="IDiscreteActionMask.GetMask"/>
+        public bool[] GetMask()
+        {
+#if DEBUG
+            if (m_CurrentMask != null)
+            {
+                AssertMask();
+            }
+#endif
+            return m_CurrentMask;
+        }
+
+        /// <summary>
+        /// Makes sure that the current mask is usable.
+        /// </summary>
+#if DEBUG
```

Comment:
bring the `#if DEBUG` in the function def too? Or bring the docstring in the `#if` - I'm not sure what XmlDoc will make of this